### PR TITLE
Refactor FeedbackQuestionsLogic.java #6435

### LINKED ISSUE #6435: Refactor FeedbackQuestionsLogic.java
1. Split large methods into several smaller methods to make the individual steps easier to understand.

2. If a particular logic is used across several methods (or used more than once in the same method), extract it into a single reusable method.

3. Add comments or rename methods where necessary to help better understand what each method is doing.

### PR DESCRIPTION
**Pull request by ablyx**
Fixes #6435 

Ready for review

### CODE REVIEW

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 1]
```diff
@@ -533,31 +538,24 @@ public void updateFeedbackQuestionNumber(FeedbackQuestionAttributes newQuestion)
      */
     private void adjustQuestionNumbers(int oldQuestionNumber,
             int newQuestionNumber, List<FeedbackQuestionAttributes> questions) {
-        
-        if (oldQuestionNumber > newQuestionNumber && oldQuestionNumber >= 1) {
-            for (int i = oldQuestionNumber - 1; i >= newQuestionNumber; i--) {
-                FeedbackQuestionAttributes question = questions.get(i - 1);
-                question.questionNumber += 1;
-                try {
+        try {
```
Comment (ablyx) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Originally, there were 2 try blocks for each `updateFeedbackQuestionWithoutResponseRateUpdate(question);` , I abstracted those 2 out to this big block because they are trying the same thing anyway. Should I undo this change?

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 2]
```diff
@@ -438,8 +428,23 @@ public int compare(FeedbackQuestionAttributes q1, FeedbackQuestionAttributes q2)
         }
         return recipients;
     }
+
+    private String getGiverTeam(String giver, InstructorAttributes instructorGiver,
```
Comment (ablyx) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: I'll change it to defaultTeam

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java (hunk 1 above)
Comment (chowyb) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Looking at the code, it might be better to use a separate method which contains just the try and catch blocks in the original code, and have both cases use that method. This keeps the amount of code in the try block to a minimum (thereby avoiding catching exceptions from other parts of the code unintentionally), while avoiding the need to duplicate code.

File: src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java [hunk 3]
```diff
@@ -287,7 +287,7 @@ public boolean isResponseVisibleTo(FeedbackParticipantType userType) {
      * @param newAttributes
      * @return
      */
-    public boolean isChangesRequiresResponseDeletion(FeedbackQuestionAttributes newAttributes) {
+    public boolean doChangesRequireResponseDeletion(FeedbackQuestionAttributes newAttributes) {
```
Comment (chowyb) [inline src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java]: Unsure if this name is okay (tag @damithc for confirmation) as `do` is not part of the current [boolean naming conventions](https://oss-generic.github.io/process/codingStandards/CodingStandard-Java.html#specific-naming-conventions).

In the event that `do` is not acceptable, it might be possible to rewrite the method name as `areResponseDeletionsRequiredForChanges` instead.

Comment (damithc) [inline src/main/java/teammates/common/datatransfer/FeedbackQuestionAttributes.java]: While it sounds 'boolean' when used e.g. `if(doChangesRequiresResponseDeletetion())`, 'do' can be interpreted as a verb. `isResponseDeletionRequiredForChanges()`?

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 4]
```diff
@@ -438,8 +428,23 @@ public int compare(FeedbackQuestionAttributes q1, FeedbackQuestionAttributes q2)
         }
         return recipients;
     }
+
+    private String getGiverTeam(String defaultTeam, InstructorAttributes instructorGiver,
+            StudentAttributes studentGiver) {
+        String giverTeam;
```
Comment (chowyb) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Would prefer this variable to be explicitly initialised to `null`

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 5]
```diff
@@ -438,8 +428,23 @@ public int compare(FeedbackQuestionAttributes q1, FeedbackQuestionAttributes q2)
...
+        } else if (isInstructorGiver) {
+            giverTeam = Const.USER_TEAM_FOR_INSTRUCTOR;
+        } else {
+            giverTeam = defaultTeam;
+        }
+        return giverTeam;
+    }
     
-    public boolean isQuestionHasResponses(String feedbackQuestionId) {
+    public boolean doesQuestionHaveResponses(String feedbackQuestionId) {
```
Comment (chowyb) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Rename this method to fit naming conventions as well.

File: src/main/java/teammates/logic/api/Logic.java [hunk 6]
```diff
@@ -1592,7 +1592,7 @@ public void deleteFeedbackQuestion(String questionId) {
      */
     
     public boolean isQuestionHasResponses(String feedbackQuestionId) {
-        return feedbackQuestionsLogic.isQuestionHasResponses(feedbackQuestionId);
+        return feedbackQuestionsLogic.doesQuestionHaveResponses(feedbackQuestionId);
```
Comment (chowyb) [inline src/main/java/teammates/logic/api/Logic.java]: You may want to rename the method in this class to be the same as that in `feedbackQuestionsLogic` as well.

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 7]
```diff
@@ -488,7 +493,7 @@ public boolean isQuestionFullyAnsweredByTeam(FeedbackQuestionAttributes question
                 }
             }
         }
-        return numberOfResponsesNeeded <= 0 ? true : false;
+        return numberOfResponsesNeeded <= 0;
```
Comment (chowyb) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Just noticed that the meaning of `numberOfResponsesNeeded` in this method is different from that in the method above (ie. in `isQuestionFullyAnsweredByUser`). You might want to consider changing this method so that the meaning of this variable is consistent with that of the previous one.

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 8]
```diff
@@ -438,8 +428,23 @@ public int compare(FeedbackQuestionAttributes q1, FeedbackQuestionAttributes q2)
...
+        String giverTeam = null;
+        boolean isStudentGiver = studentGiver != null;
+        boolean isInstructorGiver = instructorGiver != null;
+        if (isStudentGiver) {
+            giverTeam = studentGiver.team;
+        } else if (isInstructorGiver) {
+            giverTeam = Const.USER_TEAM_FOR_INSTRUCTOR;
+        } else {
+            giverTeam = defaultTeam;
+        }
```
Comment (chowyb) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Sorry I missed this earlier, but since this else statement sets a default `giverTeam` anyway, we could just shift this to the initialisation of `giverTeam`, and remove the need for this else statement.

File: src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java [hunk 9]
```diff
@@ -571,10 +563,14 @@ private void adjustQuestionNumbers(int oldQuestionNumber,
...
-
-        updateFeedbackQuestion(newAttributes, false);
+    private void updateFeedbackQuestionWithoutResponseRateUpdate(FeedbackQuestionAttributes newAttributes) {
+        try {
+            updateFeedbackQuestion(newAttributes, false);
+        } catch (InvalidParametersException e) {
+            Assumption.fail("Invalid question.");
+        } catch (EntityDoesNotExistException e) {
+            Assumption.fail("Question disappeared.");
+        }
```
Comment (chowyb) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Are there any other calls to this method other than the two above? If those two are the only calls then this should be fine, otherwise you might have to modify the other call sites to avoid redundant catch blocks.

Comment (ablyx) [inline src/main/java/teammates/logic/core/FeedbackQuestionsLogic.java]: Nope, only 2 methods call this method.
